new windows module, win_audit_rule

nwsparks opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
New module PR for win_audit_rule

Allows management of auditing rules on Windows objects (add, modify, remove)
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
win_audit_rule

nwsparks committed:
added win_audit_rule with integration test

mattclay commented:
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/37813/12/tests

nwsparks commented:
fixed now. ready_for_review

dagwieers commented on lib/ansible/modules/windows/win_audit_rule.ps1:
```diff
@@ -0,0 +1,181 @@
+
+$result = @{
+    changed = $false
+    identity_reference = $identity_reference
```
So I don't see the benefit of returning the same input as return values. The user has these values already so there is no need to return ito back to them. Worse even, it makes it harder to find the real interesting return values.

Please only return new/modified useful information.

PS Yes, older modules also did this, but we don't want to make the same mistakes for newer modules.

dagwieers commented on lib/ansible/modules/windows/win_audit_rule.ps1:
```diff
@@ -0,0 +1,181 @@
+}
+
+#Make sure target path is valid
+If (-not (Test-Path $path) )
```
We tend to add option names to cmdlets, like:
```powershell
Test-Path -Path $path
```

dagwieers commented on lib/ansible/modules/windows/win_audit_rule.ps1:
```diff
@@ -0,0 +1,181 @@
+#Make sure target path is valid
+If (-not (Test-Path $path) )
+{
+    Fail-Json $result "defined path ($path) is not found/invalid"
```
Same here:
```powershell
Fail-Json -obj $result -message "defined path '$path' is not found or is invalid"
```

dagwieers commented on lib/ansible/modules/windows/win_audit_rule.ps1:
```diff
@@ -0,0 +1,181 @@
+
+$result.current_audit_rules = Get-CurrentAuditRules $path
+$result.changed = $true
+exit-json $result
```
I would go here as well for:
```powershell
Exit-Json -obj $result
```

dagwieers commented on lib/ansible/modules/windows/win_audit_rule.ps1:
```diff
@@ -0,0 +1,181 @@
+    If (-Not $ToRemove)
+    {
+        $result.current_audit_rules = Get-CurrentAuditRules $path
+        Exit-Json $result
```
Same here:
```powershell
Exit-Json -obj $result
```

dagwieers commented on lib/ansible/modules/windows/win_audit_rule.py:
```diff
@@ -0,0 +1,139 @@
+      - For absent, only path, identity_reference, and state are required
+    required: true
+    default: present
+    choices: [ Present, Absent ]
```
Please use lower-case for these options. YAML is case-sensitive, so it will appear in the docs as Title-case.

nwsparks replied:
Thanks for the review. Do you want this done for all the options or only present/absent?

nwsparks replied:
and if all the options, how should something like NoPropagateInherit appear? I was trying to just leave it like that so I could directly pass it rather than having to convert something like no_propagate_inherit

dagwieers replied:
Well, I am mostly concerned with absent/present because that is a generic option used by other modules. Besides your examples use it as lowercase anyway. For the others, if the casing is important, leave as-is, but ensure you validate using strict casing as well, or users with casing-issues may get an ugly exception.

dagwieers commented:
I am not in favor of adding a special test_ module for testing. Use **assert** for verifying if the module works correctly, and test idempotency and check-mode support to ensure these policies are correctly created and removed. We don't want to end up in the situation where a bug in the test module negates bugs in the module.

For every operation (add, modify remove) you would usually do:
- Do it once:
  - Add using check-mode
  - Add in normal mode
  - Assert that both make a change

- Do it again:
  - Add using check-mode again
  - Add in normal mode again
  - Assert that both did not make a change

nwsparks commented:
@dagwieers for the testing module. i wasn't sure how to approach. in the primary module (win_audit_rule) i'm returning the current auditing rules to the $results object before exiting...initially I was just going to test against that, but it was suggested that out of band testing would be better.

The code to retrieve the rules was a bit much for an inline win_shell option so I did the module to retrieve them. I figured since I was already doing a comparison in the testing module I could just use its results rather than asserts in the test.

I guess my question is, what is the most preferable way of approaching this? Do you have issue with just using the module in general no matter how it is used? Or do you have issue with the comparisons to check expected results occurring withing the testing module rather than in the .yml file? Do you prefer I just compare against my returned $results from the main module?

jborean93 commented:
I don't mind having a test module to get the results of the stat instead of relying on crazy shell calls each time you want to assert against something. @dagwieers are your concerns around using the test module to check the assumptions or a test module to get results that we then assert on?

dagwieers commented:
So we discussed this during the Windows Working Group meeting, and it seems my unease with the test module is unwarranted. So forget anything I said about it :-)

nwsparks commented:
Awesome, that makes my life easier :) thanks @dagwieers 

I just pushed a commit that I believe fixed the other issues you pointed out.

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.ps1:
```diff
@@ -0,0 +1,182 @@
+$check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "bool" -default $false
+
+# these are your module parameters
+$path = Get-AnsibleParam -obj $params -name "path" -type "str" -failifempty $true -aliases "destination","dest"
```
There is a `path` type that automatically expands environment variables which would be better to use here.

nwsparks replied:
Thanks for the in depth review. I should have better investigated the type options. I'll try and work through all of these items tomorrow and validate functionality with the type changes.

jborean93 replied:
No worries, I see a lot of similarities to win_acl which is an older module and doesn't really have any of these features. Let me know once you are ready for another review and I'll do some actual testing once you're done.

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.ps1:
```diff
@@ -0,0 +1,182 @@
+    $HT = Foreach ($Obj in $ACL.Audit)
+    {
+        @{
+        identity_reference = $Obj.IdentityReference.ToString()
```
Would be best to indent these lines

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.ps1:
```diff
@@ -0,0 +1,182 @@
+# these are your module parameters
+$path = Get-AnsibleParam -obj $params -name "path" -type "str" -failifempty $true -aliases "destination","dest"
+$identity_reference = Get-AnsibleParam -obj $params -name "identity_reference" -type "str" -failifempty $true
+$rights = Get-AnsibleParam -obj $params -name "rights" -type "str"
```
Best to use the `list` type, it will automatically convert a comma separated string to a list as well as accept a list from the YAML definition as well.

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.ps1:
```diff
@@ -0,0 +1,182 @@
+$path = Get-AnsibleParam -obj $params -name "path" -type "str" -failifempty $true -aliases "destination","dest"
+$identity_reference = Get-AnsibleParam -obj $params -name "identity_reference" -type "str" -failifempty $true
+$rights = Get-AnsibleParam -obj $params -name "rights" -type "str"
+$inheritance_flags = Get-AnsibleParam -obj $params -name "inheritance_flags" -type "str" -default "ContainerInherit,ObjectInherit" -validateset 'None','ContainerInherit','ObjectInherit'
```
Does this combination work? Can you specify multiple values when `validateset` checks each value. Might be best to use the `list` type here as well.

nwsparks replied:
So...it seems to work when I have that as the default value, but not when I specify it as a parameter which I guess is why I never caught it.

nwsparks replied:
Hmm, doesn't look like -validateset can be used with the list type. I wonder if it would be worth updating it to check both default values and lists?

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.ps1:
```diff
@@ -0,0 +1,182 @@
+$rights = Get-AnsibleParam -obj $params -name "rights" -type "str"
+$inheritance_flags = Get-AnsibleParam -obj $params -name "inheritance_flags" -type "str" -default "ContainerInherit,ObjectInherit" -validateset 'None','ContainerInherit','ObjectInherit'
+$propagation_flags = Get-AnsibleParam -obj $params -name "propagation_flags" -type "str" -default "none" -ValidateSet 'InheritOnly','None','NoPropagateInherit'
+$audit_flags = Get-AnsibleParam -obj $params -name "audit_flags" -type "str" -default "success" -ValidateSet 'Success','Failure'
```
Same here, does it fire an error if you specify `audit_flags: success,failure` or is it ok with it?

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.ps1:
```diff
@@ -0,0 +1,182 @@
+If ($path -match "^HK(CC|CR|CU|LM|U):\\" -and $state -eq 'present')
+{
+    $PossibleRights = [System.Enum]::GetNames([System.Security.AccessControl.RegistryRights])
+    ($rights -split ',').Trim() | % {
```
If you use the `list` type for `$rights` this can be simplified to a foreach loop without any of the splitting or trims.

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.ps1:
```diff
@@ -0,0 +1,182 @@
+#otherwise configure acl object to add/modify user
+Else
+{
+    $flags = [System.Security.AccessControl.AuditFlags]$audit_flags
```
Does this work if you have multiple entries?

nwsparks replied:
it seems to which is great. makes changing the type to list a lot simpler.

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.ps1:
```diff
@@ -0,0 +1,182 @@
+}
+Catch {
+    $result.current_audit_rules = Get-CurrentAuditRules $path
+    Fail-Json -message "Failed to apply audit change: $($_.Exception.Message)" $result
```
Are you able to keep this similar to the other calls, e.g. `Fail-Json -obj $result -message ""`

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.py:
```diff
@@ -0,0 +1,139 @@
+        FileSystemRights U(https://msdn.microsoft.com/en-us/library/system.security.accesscontrol.filesystemrights=.aspx).
+      - If C(path) is a registry key, rights can be any right under MSDN
+        RegistryRights U(https://msdn.microsoft.com/en-us/library/system.security.accesscontrol.registryrights.aspx).
+    required: true
```
This isn't needed if you have set a default

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.py:
```diff
@@ -0,0 +1,139 @@
+      - Propagation flag on the audit rules.
+      - For more information on the choices see MSDN PropagationFlags enumeration
+        at U(https://msdn.microsoft.com/en-us/library/system.security.accesscontrol.propagationflags.aspx).
+    required: true
```
This isn't needed if you have set a default.

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.py:
```diff
@@ -0,0 +1,139 @@
+      - For more information on the choices see MSDN PropagationFlags enumeration
+        at U(https://msdn.microsoft.com/en-us/library/system.security.accesscontrol.propagationflags.aspx).
+    required: true
+    default: None
```
I believe you will need to quote this as YAML may interpret it as null.

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.py:
```diff
@@ -0,0 +1,139 @@
+    description:
+      - Defines whether to log on failure, success, or both.
+      - To log both define as comma seperated list "Success, Failure"
+    required: true
```
No needed if you have a default.

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.py:
```diff
@@ -0,0 +1,139 @@
+    description:
+      - Whether the rule should be present or absent
+      - For absent, only path, identity_reference, and state are required
+    required: true
```
Not needed if you have a default.

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.py:
```diff
@@ -0,0 +1,139 @@
+    choices: [ present, absent ]
+'''
+
+EXAMPLES = r'''
```
Would be good to have a registry example

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.py:
```diff
@@ -0,0 +1,139 @@
+      "propagation_flags": "None",
+      "rights": "Delete"
+    }
+identity_reference:
```
You will need to remove these and below as we are moving away from setting the invocation parameters.

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.py:
```diff
@@ -0,0 +1,139 @@
+    - The current rules on the defined c(path)
+    - Will return null if no rules present
+  returned: always
+  type: dictionary
```
Is this a list or a dictionary?

nwsparks replied:
I thought it is a dictionary. You're talking about this?
`
    {
      "audit_flags": "Success",
      "identity_reference": "Everyone",
      "inheritance_flags": "False",
      "is_inherited": "False",
      "propagation_flags": "None",
      "rights": "Delete"
    }
`

jborean93 commented on test/integration/targets/win_audit_rule/tasks/main.yml:
```diff
@@ -0,0 +1,399 @@
+#   register: old_os_check
+#   failed_when: old_os_check.msg != 'The win_power_plan Ansible module is only available on Server 2008r2 (6.1) and newer'
+
+- set_fact:
```
I would move these to a default file and prefix with `test_audit_rights` so they stay separate from other tests. Also things like the rights and identities set probably just best off people set in the task itself instead of a permission. Makes it easier to see what is actually happening.

nwsparks replied:
I kinda wanted to keep them as variables so that it would be easier to change the testing parameters. I think it's still pretty easy to reference by just checking the defaults file. If you really want it changed I can. I implemented the other suggestions though.

jborean93 commented on test/integration/targets/win_audit_rule/tasks/main.yml:
```diff
@@ -0,0 +1,399 @@
+    path: "{{ test_registry }}"
+
+- block:
+######################
```
Not mandatory to do but just a suggestion, you can move all the tasks in this block to another file and just put `include_tasks: tests.yml` here so you don't have a massive indented block when reading it.

jborean93 reviewed (changes requested):
Thanks for submitting the module, looks like it will be great one to have. I've left a few comments inline and once we get through them I'll put it through the paces and do some further testing.

nwsparks commented:
@jborean93 ok all the changes should be implemented

jborean93 commented:
Cool, will test it out today.

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.ps1:
```diff
@@ -0,0 +1,194 @@
+
+# these are your module parameters
+$path = Get-AnsibleParam -obj $params -name "path" -type "path" -failifempty $true -aliases "destination","dest"
+$identity_reference = Get-AnsibleParam -obj $params -name "identity_reference" -type "str" -failifempty $true
```
I would change this to just `user` to keep it in line with the `win_acl` module.

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.py:
```diff
@@ -0,0 +1,116 @@
+    description:
+      - Comma seperated list of the rights desired. Only required for adding a rule.
+      - If C(path) is a file or directory, rights can be any right under MSDN
+        FileSystemRights U(https://msdn.microsoft.com/en-us/library/system.security.accesscontrol.filesystemrights(v=vs.110).aspx).
```
the brackets in the URL itself cause issues when rendering this, just remove the `(v=vs.110)` and it will still work

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.py:
```diff
@@ -0,0 +1,116 @@
+      - If C(path) is a file or directory, rights can be any right under MSDN
+        FileSystemRights U(https://msdn.microsoft.com/en-us/library/system.security.accesscontrol.filesystemrights(v=vs.110).aspx).
+      - If C(path) is a registry key, rights can be any right under MSDN
+        RegistryRights U(https://msdn.microsoft.com/en-us/library/system.security.accesscontrol.registryrights(v=vs.110).aspx).
```
brackets in the URL

jborean93 commented on lib/ansible/modules/windows/win_audit_rule.py:
```diff
@@ -0,0 +1,116 @@
+    description:
+      - Defines what objects inside of a folder will inherit the settings.
+      - For more information on the choices see MSDN PropagationFlags enumeration
+        at U(https://msdn.microsoft.com/en-us/library/system.security.accesscontrol.inheritanceflags(v=vs.110).aspx).
```
brackets in the URL

jborean93 reviewed (changes requested):
Probably the only thing that would be good to change is `identity_reference` to `user` as it aligns with what `win_acl` is using. One issue I had when trying to set a rule on a file was

```
fatal: [Server2016]: FAILED! => {"changed": false, "current_audit_rules": "No audit rules defined on C:\\temp\\test.txt", "failed": true, "msg": "Failed to set the audit rule: Exception calling \"SetAuditRule\" with \"1\" argument(s): \"No flags can be set.\r\nParameter name: inheritanceFlags\""}
```

The task I used to do this was

```
- win_audit_rule:
    path: C:\temp\test.txt
    identity_reference: Everyone
    rights: Write,Delete
    audit_flags: Success,Failure
    state: present
```

Am I doing something wrong here?

nwsparks commented:
@jborean93 thanks. Updates the suggested changes.

It failed because $inheritance_flags had to be set to none for a file. I think setting audit rules for a specific file is something of an edge case so I like leaving the default value there to automatically make sub-objects of a folder inherit....but, I think that obviously now it needs to be addressed in some way as well as mentioned in the docs. 

Do you have a preference between these 2 options?

1. Throw an error if target is not a container and inheritance_flags is not equal to none
2. Automatically fix it for them and write a warning

I'll update integration testing to target single files as well.

nwsparks committed:
Updated integration testing to target files as well as directories
and registry keys. Split testing files apart to be more organized.

Updated powershell for better handling when targetting file objects
and optimized a bit. Removed duplicated sections that got there from a
previous merge I think.

nwsparks committed:
Decided to make all the fact names the same in integration testing.
Seemed like there would be less change of accidentally using the wrong
variable when copy/pasting that way, and not much upside to having
unique names.

Did final cleanup and fixed a few errors in the integration testing.

nwsparks commented:
@jborean93 went ahead and just made it automatically set it to 'none' if the path type is a file and made the other requested changes as well. Also added individual file testing to integration. LMK if you spot anything else.

nwsparks committed:
Fixed a bug where results was displaying a wrong value

Fixed a bug where removal was failing if multiple rules existed due to
inheritance from higher level objects.

nwsparks committed:
Resolved issue with unhandled error when used didn't have permissions
for get-acl.

Changed from setauditrule to addauditrule, see comment in script for reasoning.

Fixed state absent to be able to remove multiple entries if they exist.

nwsparks committed:
fixed docs issue

jborean93 commented:
I know you have already made the change but I'm too much a fan of automatically setting values based on the type. It would be preferable to check if the None flag is set when it is a file and throw an error. It would be good to at least document an example of how to do this as well.

nwsparks committed:
updated to fail if invalid inheritance_rule when defining a file rather than warn

nwsparks commented:
No worries. It's been updated to fail rather than warn and docs are updated as well. ready_for_review

jborean93 commented:
Thanks for this @nwsparks, when you get a chance , are you able to update the CHANGELOG.md and add this to the list of new modules for 2.5.
